source/slang/slang-ir-lower-tuple-types.cpp

```diff
@@ -0,0 +1,197 @@
+        }
+
+        LoweredTupleInfo* getLoweredTupleType(IRBuilder* builder, IRInst* type)
+        {
+            if (auto loweredInfo = loweredTuples.TryGetValue(type))
+                return loweredInfo->Ptr();
+            if (auto loweredInfo = mapLoweredStructToTupleInfo.TryGetValue(type))
+                return loweredInfo->Ptr();
+
+            if (!type)
+                return nullptr;
+            if (type->op != kIROp_TupleType)
+                return nullptr;
+
+            RefPtr<LoweredTupleInfo> info = new LoweredTupleInfo();
+            info->tupleType = (IRType*)type;
+            auto structType = builder->createStructType();
+            info->structType = structType;
+            StringBuilder nameSb, fieldNameSb;
+            nameSb << "Tuple";
+            for (UInt i = 0; i < type->getOperandCount(); i++)
+            {
+                auto elementType = maybeLowerTupleType(builder, (IRType*)(type->getOperand(i)));
+                nameSb << "_";
+                appendTypeName(nameSb, elementType);
+                auto key = builder->createStructKey();
+                fieldNameSb.Clear();
+                fieldNameSb << "value" << i;
+                builder->addNameHintDecoration(key, fieldNameSb.getUnownedSlice());
+                auto field = builder->createStructField(structType, key, (IRType*)elementType);
+                info->fields.add(field);
+            }
+            builder->addNameHintDecoration(structType, nameSb.getUnownedSlice());
+            mapLoweredStructToTupleInfo[structType] = info;
+            loweredTuples[type] = info;
+            return info.Ptr();
+        }
+
+        void addToWorkList(
+            IRInst* inst)
+        {
+            for (auto ii = inst->getParent(); ii; ii = ii->getParent())
+            {
+                if (as<IRGeneric>(ii))
+                    return;
+            }
+
+            if (workListSet.Contains(inst))
+                return;
+
+            workList.add(inst);
+            workListSet.Add(inst);
+        }
+
+        void processMakeTuple(IRMakeTuple* inst)
+        {
+            IRBuilder builderStorage;
+            auto builder = &builderStorage;
+            builder->sharedBuilder = &sharedBuilderStorage;
+            builder->setInsertBefore(inst);
+
+            auto info = getLoweredTupleType(builder, inst->getDataType());
+            auto var = builder->emitVar(info->structType);
+            for (Index i = 0; i < info->fields.getCount(); i++)
+            {
+                SLANG_ASSERT(i < (Index)inst->getOperandCount());
+                auto ptrType = builder->getPtrType(info->fields[i]->getFieldType());
+                auto addr = builder->emitFieldAddress(ptrType, var, info->fields[i]->getKey());
+                builder->emitStore(addr, inst->getOperand((UInt)i));
+            }
+            auto load = builder->emitLoad(var);
+            inst->replaceUsesWith(load);
+            inst->removeAndDeallocate();
+        }
```

Comment:
I'm a bit confused about why the `makeTuple` operation doesn't translate into a `makeStruct` instruction?

Is the answer that we don't have `makeStruct` (possible), or that there would be a concern over the order of operands (since if we *do* have a `makeStruct` I suspect it doesn't include the field keys as operands)?

Generating a new read/write temporary instead of an SSA value puts a burden on subsequent passes to undo the memory load/store operations before further optimizations become possible.

Reply:
Yeah we do have a `makeStruct`, I should probably use it.

Reply:
The code you have right now works, so I wouldn't push hard to change it unless we run into issues.